`EditorHelpSearch` improvements

AThousandShips opened the pull request:
<!--
Please target the `master` branch in priority.
PRs can target `3.x` if the same change was done in `master`, or is not relevant there.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.
You can mention in the description if the change is compatible with `3.x`.
-->
Makes editor help search show constructors/methods/signals/etc. when search is empty

This also adds operators and constructors as they were lost after #53452, though only builtins have those so it's relatively minor.

~Depends on #77554 and #77471, which it includes, please add notes for those (the first two commits) there~

* Fixes #62508

_Production edit: closes godotengine/godot-roadmap#19_

AThousandShips commented:
This PR currently:
1. Unifies search length, classes and class members are searched for with terms of length > 1
2. Makes searches of length < 2 show all the requested entries
3. Prevents searches from updating when searches go from 0->1 and 1->0 length (unless parameters change)
4. Adds constructors back to search (still needs some tweaking) and adds operators (ditto)
5. Prevents a crash when the dialog closes when searches are in progress (rarely if ever occurred before as searches were far faster)
6. Tweaks time in each search iteration to prevent perceptible slowdown (still needs some tweaking)

I would appreciate some input on the following:
- Should constructors and operators even show up (they are only available for variants, and are currently not included without this PR)
  - If yes, should they have unique icons or use the MemberMethod icon, I can try my hand at designing new ones but if someone would like to help with that I'd appreciate it and I'll add as co-author
  - And should constructors/operators list their parameters in the search (currently they are shown as tooltips) and show each or should they be unified with a single entry for each unique constructor/operator
- Should searches be unified to length one or two, I am going for length two (as I can try and solve some problems like point 3 and won't be a problem if that ends up not used, and as I think it's the most reasonable)
- Should ongoing searches disable the tree and show progress

I'm currently working on preventing search from updating when no change would occur, and ensuring nothing weird happens when the popup opens again, and tweaking the iteration to make it feel smooth

Any and all input is very welcome, this is my first PR and I just want to say that I've been following Godot and making things with it for some time now and I've rarely been in a more welcoming and productive environment, much love!

akien-mga commented on editor/editor_help_search.cpp:
```diff
@@ -401,8 +401,46 @@ bool EditorHelpSearch::Runner::_phase_match_classes() {
+		// Otherwise if term is empty add all requested members
+		else if (term.is_empty()) {
+			if (search_flags & SEARCH_CONSTRUCTORS) {
+				for (int i = 0; i < class_doc.constructors.size(); i++) {
+					match.constructors.push_back(const_cast<DocData::MethodDoc *>(&class_doc.constructors[i]));
+				}
```
I wonder if there would be a more concise way to create a `Vector<T *>` from a `Vector<T>` instead of doing it element by element. I don't have anything specific in mind, just mentioning it in case someone has an idea.

akien-mga commented on editor/editor_help_search.cpp:
```diff
@@ -401,8 +401,46 @@ bool EditorHelpSearch::Runner::_phase_match_classes() {
 			}
-			matches[class_doc.name] = match;
 		}
+		// Otherwise if term is empty add all requested members
```
We're trying to standardize comment style around having them also end sentences with a dot.

AThousandShips replied:
Thank you! Will do

akien-mga commented on editor/editor_help_search.cpp:
```diff
@@ -587,7 +622,39 @@ TreeItem *EditorHelpSearch::Runner::_create_method_item(TreeItem *p_parent, cons
 	}
 	tooltip += ")";
-	return _create_member_item(p_parent, p_class_doc->name, "MemberMethod", p_doc->name, p_text, TTRC("Method"), "method", tooltip);
+	return _create_member_item(p_parent, p_class_doc->name, "MemberConstructor", p_doc->name, p_doc->name, TTRC("Constructor"), "constructor", tooltip);
```
Not directly linked to this PR but noting it down as I review the code, this doesn't need to use `TTRC` here + `TTRGET` in `_create_member_item`. Instead it could use `TTR()` here directly and just use the translated `p_type` in `_create_member_item`.

`TTRC` is used for cases where we don't want to string translated automatically, but it's just a marker for our strings extractor to get the source string - which is then translated at runtime with `TTRGET`. While `TTR` just does the translation in place, which seems suitable here (I don't see code that expects `p_type` to be the original English string).

AThousandShips replied:
Should I change this or leave it to a dedicated PR?

akien-mga commented on editor/editor_help_search.cpp:
```diff
@@ -124,7 +145,7 @@ void EditorHelpSearch::_notification(int p_what) {
 			// Update background search.
 			if (search.is_valid()) {
-				if (search->work()) {
+				if (search->work(1000)) {
```
Worth noting that this seems to be the only call for `work()`, so if overriding the default value for its parameter, I question whether the parameter should have a default value in the first place (or if it should be overridden there too).

But either way we might need a comment to justify the magic number. IIUC this reduces the time allowed on each iteration to 1 ms - while the typical framerate with VSync should be 60 FPS so 16.6 ms. Isn't this significantly slowing down the search, forcing it to run over more `_process` iterations than it needs to?
The previous default of 100000 µs = 100 ms was likely too much indeed though, as it's the time for an average of 6 frames. Maybe this should be 10000 instead?

AThousandShips replied:
Thank you, I am in the process of tweaking this at the moment, will test this suggestion

akien-mga replied:
I tested this PR both as is (1000) and with 10000 instead and the latter feels much faster (in a `target=debug` build on Linux). But the original 100000 is even better, there's no visible delay.

1000 (this PR):
https://user-images.githubusercontent.com/4701338/176634518-deb25591-3152-4d55-8ff7-dbc4dacffa0c.mp4

10000 (my above suggestion, though it can likely be refined further):
https://user-images.githubusercontent.com/4701338/176634612-fc18a382-6a98-45d1-8cff-d00f69811505.mp4

100000 (original):
https://user-images.githubusercontent.com/4701338/176635148-f09c48eb-108b-4d3a-b8e4-a5d8b4cf43db.mp4

AThousandShips replied:
Thank you! The delay is specifically when search is length < 2 as there are a lot of entries

akien-mga replied:
Here's an example with a longer string:

1000 (still visible delay and flicker):

https://user-images.githubusercontent.com/4701338/176636397-06211660-88a8-4f49-80fd-06794fc12446.mp4

100000 (no delay, no flicker):

https://user-images.githubusercontent.com/4701338/176636506-34040453-7031-4c99-bb61-0241471ed691.mp4

So it seems to me that the original value was pretty good. Why did you need to change it?
Does this related to the crash fix you mention in the description?

AThousandShips replied:
What are your results like for short searches? I'm looking at changing the behavior between short and long searches and will amend in a moment.

AThousandShips replied:
The crash is not directly related but resolved by clearing the runner when popup is hidden, as when searches are long the runner will keep running after the tree has been cleared causing it to crash when trying to add children to non-existing tree items

AThousandShips replied:
Added two modes, one for short 0-1 characters, and one for longer "proper" searches, see how they feel

akien-mga replied:
> What are your results like for short searches length < 2? I'm looking at changing the behavior between short and long searches and will amend in a moment.

Indeed for searches with length < 2 it makes a big difference, 1000 faster than 100000, 10000 seems maybe fastest?

1000:

https://user-images.githubusercontent.com/4701338/176639590-3c380462-0d3c-43be-b51e-ea1c830def50.mp4

10000:

https://user-images.githubusercontent.com/4701338/176639607-15b38332-4866-4fb3-84f9-3d86d420882c.mp4

100000:

https://user-images.githubusercontent.com/4701338/176639614-fe383cdd-0877-42e1-b34d-dc8430c8b2f9.mp4

But I think trying to optimize for searches with length < 2 which are practically not searches but showing the whole index is the wrong approach. The content should never need to change for such searches, so it can simply be cached the first time the EditorHelp is used, and reused whenever the search string goes below 2. Then you don't even need to add logic to handle checking if going from 0 to 1 or 1 to 0, it would likely be fast enough to just refresh the tree with precomputed data so it doesn't matter (to be confirmed).

AThousandShips replied:
Thank you, that is a great idea, I'll see how I can resolve that later today, thank you for your help!

akien-mga commented on editor/editor_help_search.cpp:
```diff
@@ -165,11 +186,11 @@ void EditorHelpSearch::popup_dialog(const String &p_term) {
 	if (p_term.is_empty()) {
 		search_box->clear();
+
+		old_search = true;
```
Seems a bit weird that an empty term and cleared search box equals `old_search = true;`

AThousandShips replied:
This is to force a search if the popup closes and reopens with a short string, I'm currently seeing how best to resolve that

akien-mga commented on editor/editor_help_search.h:
```diff
@@ -61,7 +61,8 @@ class EditorHelpSearch : public ConfirmationDialog {
 	Tree *results_tree = nullptr;
 	bool old_search = false;
-	String old_term;
+	String old_term = "";
```
String is default initialized to be empty so this is not needed.

AThousandShips commented:
I implemented a basic index and filling from this index, it's very much a work in progress and code needs some cleaning up, some that belongs in the header is kept out of it to avoid recompiling lots of unrelated files etc. but I wanted to show what my current approach was. It probably needs some further tweaking to be smoother but it works far better I feel.

I will probably work more later today and clean it up and restructure a bunch of the setup, probably try move the filling into the runner to unify the code, but threw something together to see how to make it work.

AThousandShips commented:
Filling of the tree from the index is now moved to the runner, I've still got some details to work out and some style aspects to decide on, and add code for the filling without hierarchy which takes a little reworking from my current temporary solution, but it appears to work smoothly for me now.

Update on the changes of this PR:
- Short searches now show the requested types, using an index built on first use.
- The icons for member items are pre-loaded as they (especially in filling from the index) are loaded many times.
- Closing the popup while searching/filling is under way does not cause a crash (would rarely if ever occur before as ordinary searches were very fast)
- Constructors and operators are restored to the search/fill
- The tree can now be collapsed/uncollapsed recursively like the node dock tree does (holding shift while clicking the fold arrow)

Things still to do:
- Icons for constructors and operators (if anyone would be up for making such I'd be very grateful and give co-authorship)
- Filling without hierarchy, minor change for that but didn't focus on it at the moment.

For consideration:
- Should constructors and operators be shown, constructors were lost after constructors and operators were split out from methods
  - If yes, should they be shown uniquely, i.e. "Type (Constructors)" as they were before, and same for multiple copies of the same operator

Limitations:
- As before the fill does not show script classes, from what I can find there is no simple way of tracking if custom documentation was added and because of this I don't know any way of updating the index when such occurs, and therefore it's not possible to include them reliably

AThousandShips commented:
Testing out filling with no hierarchy makes me lean toward not showing it like that, it gets extremely cluttered and without adding a second index or reverting to using the documentation it won't be sorted alphabetically, so I made the decision to always draw the hierarchy when showing the index.

AThousandShips commented:
I've added some tentative icons for constructors and operators

AThousandShips commented:
I incorrectly thought that the help search didn't show custom classes at all, but as we are already experiencing a bug with custom classes being added breaking the help search I'm not sure what's going on with that, that would require the index to be rebuilt but it would require some signal on documentation being added, I'm not sure how that would be approached

AThousandShips commented:
Rebased and restructured this PR, still going to look over some details but I believe it is about ready, the following new changes have been done:
* Constructors and operators now list their argument types, as they are the only ones to be overloadable and I find it more helpful than just having to get these details from the tooltip
* Removed the update spinner as it was only to indicate progress during development

AThousandShips commented:
I believe things are ready, any feedback would be much appreciated but I believe things are working as they should, there is the problem of new scripts not showing up in search but that affects the search already and try as I might I haven't been able to identify where the problem is.

More specific decisions that I'd like some input on are:
* The choice to show all constructors, or just one "constructors", and if yes should they be shown as they are now with arguments
* Same for operators
* Opinions on the icons

AThousandShips commented:
Can this get a review?

YuriSizov commented:
So I was looking into some improvements to this dialog without realizing that you've already submitted this PR with a ton of them 🙃

This PR does a different thing than what I want to do, but I also arrived at a need for some sort of optimization to the entire runner logic. For one I want to optimize `TreeItem` creation (see https://github.com/godotengine/godot/pull/77446), but I also was looking into sorting improvements.

All of that, and your changes here, require some sort of hierarchy for sure, because even sorting must be done in a hierarchy-aware way. You introduced an index, and I'm trying to make sense of it, but there is quite a lot going on here. Perhaps it would be worth extracting it into a separate PR, and leave this one for the rest of the improvements? I can promise to review it as soon as it's ready (and this one too).

AThousandShips commented:
I will look into it! The index yes?

YuriSizov commented:
Yes, I think we should extract the index into its own PR, and rebase this one on top of that.

AThousandShips commented:
Ironing out some details but been able to separate the index part, might be able to push tomorrow

AThousandShips commented:
Update:
The index now respects updates to class index (see #77471)

Some cleaning up and modifications done, will update relevant comments of mine if they are outdated

AThousandShips commented:
When searching for all categories I now nest these member items under a common category, I feel this reduces clutter (and speeds up the process), added as a separate commit if it's not considered helpful as a feature

MewPurPur commented on editor/editor_help_search.cpp:
```diff
@@ -626,51 +898,77 @@ TreeItem *EditorHelpSearch::Runner::_create_class_item(TreeItem *p_parent, const
+	String text = p_class_doc->name + "(";
+	for (int i = 0; i < p_doc->arguments.size(); i++) {
+		const DocData::ArgumentDoc &arg = p_doc->arguments[i];
+		tooltip += arg.type + " " + arg.name;
```
Should this be `tooltip += arg.name + ": " + arg.type`?

AThousandShips replied:
Not in the code this is based on, see `_create_method_tooltip` above

MewPurPur commented on editor/editor_help_search.cpp:
```diff
@@ -685,27 +983,64 @@ TreeItem *EditorHelpSearch::Runner::_create_member_item(TreeItem *p_parent, cons
+		term(""),
+		search_flags(p_search_flags),
+		empty_icon(ui_service->get_theme_icon(SNAME("ArrowRight"), SNAME("EditorIcons"))),
+		member_annotation_icon(ui_service->get_theme_icon(SNAME("MemberAnnotation"), SNAME("EditorIcons"))),
```
`EditorStringName(EditorIcons)` should be used now. Likewise below.

AThousandShips committed:
`EditorHelpSearch` improvements

* Adds all member types to empty search
* Nests hierarchy adding constructors, methods, operators, etc. under a
  nested entry to reduce clutter

Co-authored-by: MewPurPur <mew.pur.pur@gmail.com>

AThousandShips commented:
Restructured and rebased, now includes #77554 and #77471

Included icons by MewPurPur

Currently contains a commit with nesting items in their own categories if searching for short string and all categories are included, to reduce clutter

Should be ready for reviews now, see the previous conversations for more details on some of the decisions, though some is outdated as the code is reworked

AThousandShips commented on editor/editor_help_search.cpp:
```diff
@@ -38,17 +38,39 @@
 	}
 
-	search = Ref<Runner>(memnew(Runner(results_tree, results_tree, &tree_cache, term, search_flags)));
 	set_process(true);
```
Will move this into the checks above when cleaning up

AThousandShips commented on editor/editor_help_search.h:
```diff
@@ -144,24 +167,28 @@ class EditorHelpSearch::Runner : public RefCounted {
 	TreeItem *_create_property_item(TreeItem *p_parent, const DocData::ClassDoc *p_class_doc, const DocData::PropertyDoc *p_doc);
 	TreeItem *_create_theme_property_item(TreeItem *p_parent, const DocData::ClassDoc *p_class_doc, const DocData::ThemeItemDoc *p_doc);
-	TreeItem *_create_member_item(TreeItem *p_parent, const String &p_class_name, const String &p_icon, const String &p_name, const String &p_text, const String &p_type, const String &p_metatype, const String &p_tooltip, bool is_deprecated, bool is_experimental);
+	TreeItem *_create_member_item(TreeItem *p_parent, const String &p_class_name, const StringName &p_icon, const String &p_name, const String &p_text, const String &p_type, const String &p_metatype, const String &p_tooltip, bool is_deprecated, bool is_experimental);
```
I originally cached the icons completely when creating the runner but using `StringName` now instead as it's very specific and we can use `SNAME` cheaply here to shave off some

AThousandShips commented:
Here are my performance tests based on https://github.com/godotengine/godot/pull/77471#issuecomment-1893923370, with these changes:

No caching, clearing tree instead of storing all the items
```
EditorHelp search done in 349790 usec

EditorHelp search done in 131627 usec
EditorHelp search done in 42886 usec
EditorHelp search done in 28333 usec
EditorHelp search done in 26360 usec
EditorHelp search done in 26840 usec
EditorHelp search done in 24004 usec
EditorHelp search done in 22215 usec
EditorHelp search done in 21156 usec
EditorHelp search done in 20433 usec

EditorHelp search done in 284755 usec

EditorHelp search done in 132012 usec
EditorHelp search done in 40711 usec
EditorHelp search done in 26778 usec
EditorHelp search done in 25513 usec
EditorHelp search done in 23497 usec
EditorHelp search done in 21685 usec
EditorHelp search done in 20848 usec
EditorHelp search done in 20610 usec

EditorHelp search done in 286064 usec
```

No caching, but not clearing tree (this builds up used objects, but don't think it necessarily affects this part of the performance)
```
EditorHelp search done in 319704 usec

EditorHelp search done in 60183 usec
EditorHelp search done in 30179 usec
EditorHelp search done in 26794 usec
EditorHelp search done in 26078 usec
EditorHelp search done in 25397 usec
EditorHelp search done in 23168 usec
EditorHelp search done in 22543 usec
EditorHelp search done in 21628 usec
EditorHelp search done in 21260 usec

EditorHelp search done in 289458 usec

EditorHelp search done in 62680 usec
EditorHelp search done in 29478 usec
EditorHelp search done in 26036 usec
EditorHelp search done in 25211 usec
EditorHelp search done in 22878 usec
EditorHelp search done in 22875 usec
EditorHelp search done in 21384 usec
EditorHelp search done in 21342 usec

EditorHelp search done in 284969 usec
```

With caching
```
EditorHelp search done in 320378 usec

EditorHelp search done in 50075 usec
EditorHelp search done in 29114 usec
EditorHelp search done in 26448 usec
EditorHelp search done in 25903 usec
EditorHelp search done in 25190 usec
EditorHelp search done in 23436 usec
EditorHelp search done in 21987 usec
EditorHelp search done in 21060 usec
EditorHelp search done in 21046 usec

EditorHelp search done in 97474 usec

EditorHelp search done in 49374 usec
EditorHelp search done in 28493 usec
EditorHelp search done in 26446 usec
EditorHelp search done in 25367 usec
EditorHelp search done in 24538 usec
EditorHelp search done in 22766 usec
EditorHelp search done in 22145 usec
EditorHelp search done in 21414 usec
EditorHelp search done in 20597 usec

EditorHelp search done in 96840 usec
```

YuriSizov commented:
Could you please rebase on the current master? We had an issue with macOS builds which failed your CI.

AThousandShips commented:
Will do was waiting for it to get merged, thank you for notifying me 🙂

Mickeon commented on editor/editor_help_search.cpp:
```diff
@@ -394,6 +419,226 @@ bool EditorHelpSearch::Runner::_is_class_disabled_by_feature_profile(const Strin
+		return false;
+	}
+
+	if (!iterator_stack[iterator_stack.size() - 1]) {
```
Weird? I am not sure what this is for at a glance, nor the order of operations here.

AThousandShips replied:
It checks if the last entry in the stack has data, it's a standard operation

Mickeon commented on editor/editor_help_search.cpp:
```diff
@@ -394,6 +419,226 @@ bool EditorHelpSearch::Runner::_is_class_disabled_by_feature_profile(const Strin
+		DocData::ClassDoc *class_doc = EditorHelp::get_doc_data()->class_list.getptr(iterator_stack[iterator_stack.size() - 1]->get());
+
+		iterator_stack[iterator_stack.size() - 1] = iterator_stack[iterator_stack.size() - 1]->prev();
+		if (!iterator_stack[iterator_stack.size() - 1]) {
```
Ditto?
Also repeating `size - 1` every time is a bit of a hassle, but I understand why it is done.

AThousandShips replied:
It indeed has to be done as it isn't constant

Mickeon commented on editor/editor_help_search.cpp:
```diff
@@ -394,6 +419,226 @@ bool EditorHelpSearch::Runner::_is_class_disabled_by_feature_profile(const Strin
+		TreeItem *item = _create_class_hierarchy(class_doc, !(search_flags & SEARCH_CLASSES));
+
+		// If class has no child classes, fold it.
+		if (!item->get_first_child()) {
+			item->set_collapsed(true);
+		} else {
+			item->set_collapsed(false);
+		}
```
If a comment needs to explain what's happening here anyway, this could be compressed into a single line, I suppose?
```suggestion
		item->set_collapsed(!item->get_first_child());
```
Although the comment itself could be more descriptive (e.g. "If the class has inheriting classes, fold its item")

AThousandShips replied:
It's less clear IMO as it's not a boolean value, but might change that

> If the class has inheriting classes, fold its item

You mean "If the class has no inheriting classes, fold its item"?

Mickeon replied:
Mhm!

Mickeon commented on editor/editor_help_search.cpp:
```diff
@@ -394,6 +419,226 @@ bool EditorHelpSearch::Runner::_is_class_disabled_by_feature_profile(const Strin
+		}
+
+		// Create common header if required.
+		const bool search_all = (search_flags & SEARCH_ALL) == SEARCH_ALL;
```
Two can play at this game.
```suggestion
		bool search_all = (search_flags & SEARCH_ALL) == SEARCH_ALL;
```

AThousandShips replied:
Ditto

AThousandShips commented on editor/editor_help_search.cpp:
```diff
@@ -394,6 +419,226 @@ bool EditorHelpSearch::Runner::_is_class_disabled_by_feature_profile(const Strin
+		}
+
+		// If class matches the flags, add it to the matched stack.
+		const bool class_matched =
```
That's a local variable, we don't enforce const or non const for that

AThousandShips commented on editor/editor_help_search.cpp:
```diff
@@ -895,52 +1153,83 @@ TreeItem *EditorHelpSearch::Runner::_create_class_item(TreeItem *p_parent, const
+TreeItem *EditorHelpSearch::Runner::_create_method_item(TreeItem *p_parent, const DocData::ClassDoc *p_class_doc, const MemberMatch<DocData::MethodDoc> &p_match) {
 	String tooltip = _build_method_tooltip(p_class_doc, p_match.doc);
-	return _create_member_item(p_parent, p_class_doc->name, "MemberMethod", p_match.doc->name, p_text, TTRC("Method"), "method", tooltip, p_match.doc->keywords, p_match.doc->is_deprecated, p_match.doc->is_experimental, p_match.name ? String() : p_match.keyword);
+	return _create_member_item(p_parent, p_class_doc->name, SNAME("MemberMethod"), p_match.doc->name, p_match.doc->name, TTRC("Method"), "method", tooltip, p_match.doc->keywords, p_match.doc->is_deprecated, p_match.doc->is_experimental, p_match.name ? String() : p_match.keyword);
```
That's from the old code, not my addition, can look at it later though

AThousandShips commented on editor/editor_help_search.cpp:
```diff
@@ -880,6 +1137,7 @@ TreeItem *EditorHelpSearch::Runner::_create_class_item(TreeItem *p_parent, const
 			item->add_button(0, warning_icon, 0, false, TTR("This class is marked as experimental."));
 		}
 	}
+	item->set_collapsed(false);
```
Sure, it's because it has to restore the folding state as it's cached, can add a note

AThousandShips commented on editor/editor_help_search.cpp:
```diff
@@ -697,7 +931,7 @@ bool EditorHelpSearch::Runner::_phase_select_match() {
 }
 
-void EditorHelpSearch::Runner::_match_method_name_and_push_back(Vector<DocData::MethodDoc> &p_methods, Vector<MemberMatch<DocData::MethodDoc>> *r_match_methods) {
+void EditorHelpSearch::Runner::_match_method_name_and_push_back(Vector<DocData::MethodDoc> &p_methods, LocalVector<MemberMatch<DocData::MethodDoc>> *r_match_methods) {
```
`LocalVector` doesn't do any sharing, so it's faster when you don't need to

AThousandShips commented on editor/editor_help_search.cpp:
```diff
@@ -394,6 +419,226 @@ bool EditorHelpSearch::Runner::_is_class_disabled_by_feature_profile(const Strin
+
+		// Create common header if required.
+		const bool search_all = (search_flags & SEARCH_ALL) == SEARCH_ALL;
+
```
Could use foreach maybe, but this uses addressing so it's not entirely the same, can look at that though too

Geometror reviewed (commented):
I'm not sure about the 2 character threshold. It feels weird to me since most other software I know begins filtering at one character.
I know, performance might be a problem, but just starting the search at 2 character isn't the right solution IMO.

AThousandShips commented:
That's not new, so that's a separate matter IMO

Geometror commented:
Well, you united the search term length thresholds so that classes are searched only after >=2 characters. (1 prior)

AThousandShips commented:
No it searches for classes when terms are short, check the code for the else block

AThousandShips commented:
There, now matches fully on short searches and checks keywords on short searches as well, but only checks classes and does not filter members on short searches

Geometror reviewed (approved):
Tested again and works as expected now.
Looks good, great work!

Things out of scope of this PR which came to my mind:
- It would be nice to have unit tests for this
- The icons look fine to me, although I think we should color (all of) them
- Search should be async, so there is no noticeable input delay (any thoughts on this?)

AThousandShips commented:
Don't see how we could add unit tests for this, what would we test? It's internal and we don't have unit tests for any editor features at all

The search is async but it fills in the list online, so not sure how that could be solved except with having a timer delaying starting searching until after a pause in inputting

Mickeon commented:
> The search is async but it fills in the list online, so not sure how that could be solved except with having a timer delaying starting searching until after a pause in inputting

I believe most of the times it's handled this way, yes, although I've not seen this kind of thing done in the editor as far as I know. Except, say, the script editor's autocompletion.

akien-mga commented:
It doesn't seem to work properly for me on Linux (Fedora 40 on Wayland, running X11 editor). Testing latest commit from this PR rebased on `master` (108c603f91b94100a1adc989316a372f0a6f8989).
It looks like it's not refreshing/drawing properly after I start typing.

The search with empty filter looks correct:
![image](https://github.com/user-attachments/assets/3a11d5bc-afab-4eea-baac-1cd016695307)

But after typing a search, nothing is shown:
![image](https://github.com/user-attachments/assets/0eee36ad-85a4-4f25-a125-a11930c980b7)

Reopening the dialog makes it redraw and now results are visible:
![image](https://github.com/user-attachments/assets/ba690ac2-2bc2-4ddb-869e-596c901c3a13)

AThousandShips commented:
Will investigate

AThousandShips commented:
Can confirm, recent regression no clue what might have caused it, will try digging

AThousandShips commented:
This happens on `master` as well so is a regression outside this

AThousandShips commented:
Wrote a fix for the regression, including here temporarily to help with testing, see:
* https://github.com/godotengine/godot/pull/96206

Geometror commented:
> Don't see how we could add unit tests for this, what would we test? It's internal and we don't have unit tests for any editor features at all

Yep, that is one thing that needs to improve in order to find editor regressions as soon as possible. We have the [Editor] tag for test cases which should allow testing this (if not it needs to be extended/fixed). The only thing we need is another small refactoring (maybe one or two interface methods) to make it testable. Writing some unit tests to find regressions would be easy then.

> The search is async but it fills in the list online, so not sure how that could be solved except with having a timer delaying starting searching until after a pause in inputting

Alright, maybe it's not that bad in release builds, but at some point I think we should find a solution for the lag (although not here and now).

akien-mga commented:
I don't remember why the "Nest hierarchy" commit is separate, is that still up for discussion whether this should be included or not? Is it meant to stay separate for clarity, or be squashed eventually? If kept separate, the commit message should likely be more explicit (it should be "standalone" too).

AThousandShips commented:
Will squash it if everything is approved, once CI has finished and will clarify the original commit message as well, I think the nested hierarchy (adding headings for constructors, methods, etc.) is good to keep

akien-mga commented:
Tested briefly, it looks good overall, but I'm not sure I understand what triggers showing the "methods"/"signals"/etc. headers like this:
![image](https://github.com/user-attachments/assets/9852022a-56c1-45ed-bbed-345a5e356450)

I get this by searching "c" (why is it narrowing down on "Curve" specifically btw?), but not if I search "curve".
![image](https://github.com/user-attachments/assets/86101a23-3e55-4693-ae8b-34136fca1916)

AThousandShips commented:
The showing of the sub-headers is done if there's more than one entry of that kind, with a short search only classes are filtered at the present

It picks `Curve` specifically because of the scoring system which picks short results over long results with equal score otherwise (this is unchanged)

Filtering on members on short searches could be added but I think it's of limited use as searching for just `c` is not going to find much of value (at least compared to the cost of doing so), so for short searches it just filters classes that are found and shows all their members

Adding filtering on member items would take a little more work so if we decide to add that as well I'd prefer to do it separately

AThousandShips commented:
Just gonna push a usability improvement for navigation after I've tested it

AThousandShips commented:
Thank *you*! My first ever PR finally merged! :tada:
